fem/bilinearform_ext.cpp

```diff
@@ -36,12 +139,17 @@ const Operator *BilinearFormExtension::GetRestriction() const
 
 
 // Data and methods for partially-assembled bilinear forms
-PABilinearFormExtension::PABilinearFormExtension(BilinearForm *form) :
-   BilinearFormExtension(form),
-   trialFes(a->FESpace()), testFes(a->FESpace()),
-   localX(trialFes->GetNE() * trialFes->GetFE(0)->GetDof() * trialFes->GetVDim()),
-   localY( testFes->GetNE() * testFes->GetFE(0)->GetDof() * testFes->GetVDim()),
-   elem_restrict(new ElemRestriction(*a->FESpace())) { }
+PABilinearFormExtension::PABilinearFormExtension(BilinearForm *form)
+   : BilinearFormExtension(form),
+     trialFes(a->FESpace()), testFes(a->FESpace()),
+     elem_restrict(new ElemRestriction(*a->FESpace()))
+{
+   const Table &el_dof = trialFes->GetElementToDofTable();
+   const int esize = el_dof.Size_of_connections()*trialFes->GetVDim();
+   localX.SetSize(esize, Device::GetMemoryType());
+   localY.SetSize(esize, Device::GetMemoryType());
+   localY.UseDevice(); // ensure 'localY = 0.0' is done on device
```

Comment:
It is simple: if you call `UseDevice()` for a `Vector` then all vector operations with it will be done using the `mfem::Device` configuration.

Reply:
Since this PR is about GPU port, wouldn't it make more sense that everything goes on the GPU by default? Thus we would have to tag the few vector that don't go on the GPU. This would probably also allow to omit the `Device::GetMemoryType` argument, because then when you use `DontUseDevice()` that would always mean the same thing so this argument becomes irrelevant. This might also avoid the duplication of method like `SetSize`.